Data visitors

gortiz opened the pull request:
This PR adds 3 visitors: `DataTypeVisitor`, `SingleValueVisitor` and `MultiValueVisitor`.

The [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) is well-known and heavily used in _old_ versions of Java where `switch` statement cannot be applied to classes. With the inclusion of pattern matching and switch expressions in modern Java versions the visitor pattern is not that important.

In Apache Pinot it can be useful on a lot of different object domains. In this PR I added to:
- FieldSpec.DataType. As explained in the javadoc, given that DataType is an enum we can apply a switch and we actually do it in several places, but the switch has two advantages over switch: it is typesafe and it can be composed.
- Some predicates: Here I have added the visitors in order to offer internal typesafe visibility of the predicates. The idea is to be able to extract the matching value of the predicate (for example, the actual literal value we are comparing with in a `Eq` predicate) without having to know anything about the actual predicate apart.

The later is quite useful in a proprietary plugin we are developing in StarTree.

gortiz commented:
@walterddr: I'm nominating you as reviewer because I know you like visitors ;)

walterddr reviewed (commented):
we dont have a single concrete impl of the visitors in this PR so this PR is merely just a interface addition? 

do we have a concrete usage of this in a follow up PR?

gortiz commented:
> we dont have a single concrete impl of the visitors in this PR so this PR is merely just a interface addition?

Yes, just the interface in this PR.

> do we have a concrete usage of this in a follow up PR?

I'm working on a proprietary extension and I found this pattern would help me in the implementation. I didn't dedicate time trying to look for possible usages in Apache Pinot, but I'm sure we can use these interfaces in code we already wrote. I'm using them to be able to access the `matchingValue` attribute in a safe way without having to create a new method on the class to export it and without doing boxing (although I need to create a new Visitor, so the performance change shouldn't be significant). Alternatively we could just add a `getMatchingValue` (and similar to IN, NEQ and NIN).

gortiz commented:
I've added some tests to show how to use the visitors and actually test that methods are been called. Also, I've changed the `OpRawPredicateEvaluator` (where `Op` is `Eq`, `In`, etc) classes visibility in order to be able to actually have the `accept` method accessible from outside.

walterddr reviewed (approved):
lgtm. several questions we can follow up.

- is there any reason why RegexpLike and FST-Based factories not part of the visitor pattern change?
- DataType visitor tests are not showing exactly how this will be useful yet.

walterddr commented:
CC @yupeng9 @xiangfu0 @Jackie-Jiang please kindly take another look

gortiz commented:
It doesn't seem that more people is going to review this. Can we merge it?

Jackie-Jiang commented on pinot-spi/src/main/java/org/apache/pinot/spi/data/MultiValueVisitor.java:
```diff
@@ -0,0 +1,99 @@
+
+public interface MultiValueVisitor<R> {
+
+  R visitInt(int[] value);
```
Based on my understanding of visitor, and also from this [blog](https://www.baeldung.com/java-visitor-pattern), visitor should not return anything. Currently we force the visitor to return the same type result when visiting different data types which seems kind of strange. Should we remove the return and always handle the visiting logic within the method?

walterddr replied:
IMO many visitor makes return type generic and we put `Void` type when no return type is necessary. this should be fine.

gortiz replied:
Yep, usually is better to return something in visitors. Otherwise _visitors that return_ should store some state that the caller has to consult later with a getter method. That means that the visitor must be stateful and the code will be uglier. Instead, if you let visitor to return, it will be easier to use from caller perspective and you can always use Void when not needed.

In fact, in a previous life when I heavily use visitors, I usually introduce a second generic `A` (from Argument) that was used to pass contextual information to the visitor. Visit methods were defined as `R visit(SomeClass value, A argument)`. By doing that visitor instances are almost always stateless. I didn't do that here because it is not a well common pattern and didn't want to make it too complex.

walterddr replied:
there's also a visitor return + context
- visitor return carries visiting info (related to the visit sequence) and 
- the context carries contextual info that's unrelated to the visiting sequence

But largely depending on the use case. i don't think we need to go that far here. (we went for this approach in query planner though)

Jackie-Jiang reviewed (approved):
LGTM otherwise
